qubesadmin/tools/qvm_ls.py

```diff
@@ -710,16 +786,109 @@ def main(args=None, app=None):
+        # filter only VMs with specific label(s)
+        domains_labeled = []
+        spinner.show('Filtering based on labels...')
+        for dom in domains:
+            if dom.label.name in args.label:
+                domains_labeled.append(dom)
+            spinner.update()
+        domains = domains_labeled
+        spinner.hide()
+
     if args.tags:
         # filter only VMs having at least one of the specified tags
         domains = [dom for dom in domains
                    if set(dom.tags).intersection(set(args.tags))]
 
+    if args.no_tags:
+        # exclude VMs having at least one of the specified tags
+        domains = [dom for dom in domains
+                   if not set(dom.tags).intersection(set(args.no_tags))]
+
+    if args.based_on:
+        # Filter only VMs based on specific TemplateVM
+        child_domains = []
+        spinner.show('Filtering results to VMs based on their template...')
+        for dom in domains:
+            if getattr(dom, 'template', '') in args.based_on:
+                child_domains.append(dom)
+            spinner.update()
+        domains = child_domains
+        spinner.hide()
+
+    if args.conn_netvm:
+        # Filter only VMs connecting with specific netvm
+        domains_connecting = []
+        spinner.show('Filtering results to VMs based on their netvm...')
+        for dom in domains:
+            if getattr(dom, 'netvm', '') in args.conn_netvm:
+                domains_connecting.append(dom)
+            spinner.update()
+        domains = domains_connecting
+        spinner.hide()
+
+    if args.internal in ['y', 'yes']:
+        domains = [d for d in domains if d.features.get('internal', None)
+                   in ['1', 'true', 'True']]
+    elif args.internal in ['n', 'no']:
+        domains = [d for d in domains if not d.features.get('internal', None)
+                   in ['1', 'true', 'True']]
+
+    if args.servicevm in ['y', 'yes']:
+        domains = [d for d in domains if d.features.get('servicevm', None)
+                   in ['1', 'true', 'True']]
+    elif args.servicevm in ['n', 'no']:
+        domains = [d for d in domains if not d.features.get('servicevm', None)
+                   in ['1', 'true', 'True']]
+
+    if args.pending_update:
+        domains = [d for d in domains if
+                   d.features.get('updates-available', None)]
+
+    if args.features:
+        # Filter only VMs with specified features
+        for feature in args.features:
+            try:
+                key, value = feature.split('=', 1)
+            except ValueError:
+                parser.error("Invalid argument --features {}".format(feature))
+            if not key:
+                parser.error("Invalid argument --features {}".format(feature))
+            if value == '':
+                value = None
+            elif value in ['\'\'', '""']:
+                value = ''
+            domains = [d for d in domains if d.features.get(key, None) == value]
```

Comment:
Have you benchmarked this? Same thing about `args.prefs` later on. It may be slow, but decreasingly slow as the `domains` supposedly gets smaller.

Validating the arguments in the for loop seems okay, but the features.get() is ran for every feature instead of once per feature key pair.

If the runtime is not bad, then may be okay to leave as it is.

Reply:
> Have you benchmarked this?

They are almost instant in dom0 on this HP Elitebook 820 G1 from Windows 7 era. But noticeable in other domains (as policy checks come into play). Not painful. Very much usable. Maybe I should consider adding the spinner for features and preferences filtering as well.